ovos_local_backend/database.py

```diff
@@ -0,0 +1,951 @@
+
+    timestamp = db.Column(db.Integer, primary_key=True)  # unix seconds
+    uuid = db.Column(db.String(255))
+
+
+def add_metric(uuid, name, data):
+    count = db.session.query(Metric).count() + 1
+    metric_id = f"@{uuid}|{name}|{count}"
+    entry = Metric(
+        metric_id=count,
+        metric_type=name,
+        metadata_json=data,
+        uuid=uuid,
+        timestamp=time.time()
+    )
+    db.session.add(entry)
+    db.session.commit()
+    return entry
+
+
+def get_metric(metric_id):
+    return Metric.query.filter_by(metric_id=metric_id).first()
+
+
+def delete_metric(metric_id):
+    entry = get_metric(metric_id)
+    if not entry:
+        return False
+    db.session.delete(entry)
+    db.session.commit()
+    return True
+
+
+def update_metric(metric_id, data):
+    metric: Metric = get_metric(metric_id)
+    if not metric:
+        uuid, name, count = metric_id.split("|")
+        uuid = uuid.lstrip("@")
+        metric = add_metric(uuid, name, data)
+    else:
+        metric.metadata_json = data
+        db.session.commit()
+    return metric
+
+
+def list_metrics():
+    return Metric.query.all()
+
+
+def add_wakeword_definition(ww_id, name, lang, ww_config, plugin):
+    entry = WakeWordDefinition(ww_id, lang=lang, name=name,
+                               ww_config=ww_config, plugin=plugin)
+    db.session.add(entry)
+    db.session.commit()
+    return entry
+
+
+def get_wakeword_definition(ww_id):
+    return WakeWordDefinition.query.filter_by(ww_id=ww_id).first()
+
+
+def delete_wakeword_definition(ww_id):
+    entry = get_wakeword_definition(ww_id)
+    if not entry:
+        return False
+    db.session.delete(entry)
+    db.session.commit()
+    return True
+
+
+def list_wakeword_definition():
+    return WakeWordDefinition.query.all()
+
+
+def list_voice_definitions():
+    return VoiceDefinition.query.all()
+
+
+def update_wakeword_definition(ww_id, name=None, lang=None, ww_config=None, plugin=None):
+    ww_def: WakeWordDefinition = get_wakeword_definition(ww_id)
+    if not ww_def:
+        ww_def = add_wakeword_definition(ww_id=ww_id, lang=lang, name=name,
+                                         ww_config=ww_config, plugin=plugin)
+    else:
+        if name:
+            ww_def.name = name
+        if plugin:
+            ww_def.plugin = plugin
+        if lang:
+            ww_def.lang = lang
+        if ww_config:
+            ww_def.ww_config = ww_config
+        db.session.commit()
+    return ww_def
+
+
+def add_device(uuid, token, name=None, device_location="somewhere", opt_in=False,
+               location=None, lang=None, date_format=None, system_unit=None,
+               time_format=None, email=None, isolated_skills=False,
+               ww_id=None, voice_id=None):
+    lang = lang or CONFIGURATION.get("lang") or "en-us"
+
+    email = email or \
+            _mail_cfg.get("recipient") or \
+            _mail_cfg.get("smtp", {}).get("username")
+
+    loc = location or _loc
```

Comment:
this line does not work.  `location` is not an empty dict so it never gets to `_loc`

Reply:
`location['coordinate']` is the only one that throws an error, so I replaced with a check for that.
```
if not location["coordinate"]:
    loc = _loc
else:
    loc = location
```
then it worked.  Not sure how else to handle this

Reply:
im not following, where is that coming from? we certainly dont want to partially read the default loc only, its all or nothing

is something passing an incomplete location here?